### PR TITLE
update WeightedSSLModel

### PR DESCRIPTION
**Pull request by Adel-Moumen**
## What does this PR do?

This PR attempts to improve the class `WeightedSSLModel`. First I choose to move this class from the w2v2 file as it is also working for other SSL models such as Data2vec. I also did some simplification such as in the forward where we were naively iterating over the layers while we could do it in two torch operations which can significantly improve the efficiency of the class (e.g. we gains 10 seconds on the test-clean of LibriSpeech with this). Furthermore, I removed the `num_layers` parameter as the value can be retrieved in the config dict. 

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/lobes/models/huggingface_transformers/wav2vec2.py [hunk 1]
```diff
@@ -309,83 +309,3 @@ def override_config(self, config):
...
-        norm_weights = torch.nn.functional.softmax(self.weights, dim=-1)
-        # Layernorming the layers representations if asked
-        if self.layernorm:
-            hidden_states = [
-                F.layer_norm(t, (t.shape[-1],)) for t in hidden_states
-            ]
-        # Summing the weighted layers
-        weighted_feats = hidden_states[0] * norm_weights[0]
-        for i in range(1, len(hidden_states)):
-            weighted_feats += hidden_states[i] * norm_weights[i]
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/huggingface_transformers/wav2vec2.py:374]: I specifically optimise those lines in a single one : `weighted_feats = (hidden_states * norm_weights[:, None, None, None]).sum(axis=0)`.

File: speechbrain/lobes/models/huggingface_transformers/weighted_ssl.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    def __init__(self, hub, layernorm=False):
+        super().__init__(source=hub)
+        self.model.eval()
+        self.num_layers = self.config.num_hidden_layers + 1
+        # Initializing the learnable weights
+        zero_init = torch.cat([torch.zeros(self.num_layers)])
+        self.weights = torch.nn.Parameter(zero_init, requires_grad=True)
+        self.layernorm = layernorm
+
+    def forward(self, wav, wav_lens=None):
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/huggingface_transformers/weighted_ssl.py:58]: BTW, @TParcollet do you know why `wav_lens` is not being passed in the model ?

**Comment by mravanelli** [comment]
@salah-zaiem do you think you can take a look at this PR?

**Review by salah-zaiem — APPROVED** [review]
Looks good to me! I've been planning to remove the for loop for some time, nice catch!

**Comment by mravanelli** [comment]
I did a code inspection and everything LGTM! Thank you @Adel-Moumen for the nice refactor and @salah-zaiem for the review.
